server/schedule/operator/step.go

```diff
@@ -438,3 +440,331 @@ func (al AddLightLearner) Influence(opInfluence OpInfluence, region *core.Region
+		return errors.New("cannot demote leader peer")
+	}
+	return nil
+}
+
+// Influence calculates the store difference that current step makes.
+func (df DemoteFollower) Influence(opInfluence OpInfluence, region *core.RegionInfo) {}
+
+// DemoteVoter is very similar to DemoteFollower. But it allows Demote Leader.
+// Note: It is not an OpStep, only a sub step in ChangePeerV2Enter and ChangePeerV2Leave.
+type DemoteVoter struct {
+	ToStore, PeerID uint64
+}
+
+func (dv DemoteVoter) String() string {
+	return fmt.Sprintf("demote voter peer %v on store %v to learner", dv.PeerID, dv.ToStore)
+}
+
+// ConfVerChanged returns the delta value for version increased by this step.
+func (dv DemoteVoter) ConfVerChanged(region *core.RegionInfo) bool {
+	peer := region.GetStoreLearner(dv.ToStore)
+	return peer.GetId() == dv.PeerID
+}
+
+// IsFinish checks if current step is finished.
+func (dv DemoteVoter) IsFinish(region *core.RegionInfo) bool {
+	if peer := region.GetStoreLearner(dv.ToStore); peer != nil {
+		if peer.GetId() != dv.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", dv.String()), zap.Uint64("obtain-learner", peer.GetId()))
+		}
+		return peer.GetId() == dv.PeerID
+	}
+	return false
+}
+
+// ChangePeerV2Enter is an OpStep that uses joint consensus to request all PromoteLearner and DemoteVoter.
+type ChangePeerV2Enter struct {
+	PromoteLearners []PromoteLearner
+	DemoteVoters    []DemoteVoter
+}
+
+func (cpe ChangePeerV2Enter) String() string {
+	b := &strings.Builder{}
+	_, _ = b.WriteString("use joint consensus")
+	for _, pl := range cpe.PromoteLearners {
+		_, _ = fmt.Fprintf(b, ", promote learner peer %v on store %v to voter", pl.PeerID, pl.ToStore)
+	}
+	for _, dv := range cpe.DemoteVoters {
+		_, _ = fmt.Fprintf(b, ", demote voter peer %v on store %v to learner", dv.PeerID, dv.ToStore)
+	}
+	return b.String()
+}
+
+// ConfVerChanged returns the delta value for version increased by this step.
+func (cpe ChangePeerV2Enter) ConfVerChanged(region *core.RegionInfo) uint64 {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer.GetId() != pl.PeerID || !core.ReadyToBecomeVoter(peer) {
+			return 0
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStoreVoter(dv.ToStore)
+		if peer != nil && (peer.GetId() != dv.PeerID || !core.ReadyToBecomeLearner(peer)) {
+			return 0
+		}
+	}
+	return uint64(len(cpe.PromoteLearners) + len(cpe.DemoteVoters))
+}
+
+// IsFinish checks if current step is finished.
+func (cpe ChangePeerV2Enter) IsFinish(region *core.RegionInfo) bool {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer != nil && peer.GetId() != pl.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", pl.String()), zap.Uint64("obtain-voter", peer.GetId()))
+		}
+		if peer.GetId() != pl.PeerID || peer.GetRole() != metapb.PeerRole_IncomingVoter {
+			return false
+		}
```

Comment:
It seems we only need to check `peer.GetRole() != metapb.PeerRole_IncomingVoter` here. `peer.GetId() != pl.PeerID` is covered by the first `if` branch.

Reply:
The first `if` branch is just for logging, the situation is different from the second `if`.
If `peer == nil`, `peer.GetId() != pl.PeerID == true`, but `peer != nil && peer.GetId() != pl.PeerID == false`.

Reply:
got it.